lib/db_connection/connection_pool.ex

```diff
@@ -0,0 +1,428 @@
+        {:noreply, data}
+      :stop ->
+        stop_holder(holder, extra)
+        {:noreply, data}
+    end
+  end
+
+  def handle_info({:timeout, deadline, {queue, holder, pid, len}}, {_, queue, _} = data) do
+    # Check that timeout refers to current holder (and not previous)
+    try do
+      :ets.lookup_element(holder, @holder_key, 3)
+    rescue
+      ArgumentError ->
+        :ok
+    else
+      ^deadline ->
+        :ets.update_element(holder, @holder_key, {3, nil})
+        message = "client #{inspect pid} timed out because " <>
+            "it queued and checked out the connection for longer than #{len}ms"
+        err = DBConnection.ConnectionError.exception(message)
+        disconnect_holder(holder, err)
+      _ ->
+        :ok
+    end
+    {:noreply, data}
+  end
+
+  def handle_info({:timeout, poll, {time, last_sent}}, {_, _, %{poll: poll}} = data) do
+    {status, queue, codel} = data
+    # If no queue progress since last poll check queue
+    case :ets.first(queue) do
+        {sent, _, _} when sent <= last_sent and status == :busy ->
+            delay = time - sent
+            timeout(delay, time, queue, start_poll(time, sent, codel))
+        {sent, _, _} ->
+            {:noreply, {status, queue, start_poll(time, sent, codel)}}
+        _ ->
+            {:noreply, {status, queue, start_poll(time, time, codel)}}
+    end
+  end
+
+  def handle_info({:timeout, idle, {time, last_sent}}, {_, _, %{idle: idle}} = data) do
+    {status, queue, codel} = data
+    # If no queue progress since last idle check oldest connection
+    case :ets.first(queue) do
+        {sent, holder} = key when sent <= last_sent and status == :ready ->
+            :ets.delete(queue, key)
+            ping(holder, queue, start_idle(time, last_sent, codel))
+        {sent, _} ->
+            {:noreply, {status, queue, start_idle(time, sent, codel)}}
+        _ ->
+            {:noreply, {status, queue, start_idle(time, time, codel)}}
+    end
+  end
+
+  defp timeout(delay, time, queue, codel) do
+    case codel do
+      %{delay: min_delay, next: next, target: target, interval: interval}
+          when time >= next and min_delay > target ->
+        codel = %{codel | slow: true, delay: delay, next: time + interval}
+        drop_slow(time, target * 2, queue)
+        {:noreply, {:busy, queue, codel}}
+      %{next: next, interval: interval} when time >= next ->
+        codel = %{codel | slow: false, delay: delay, next: time + interval}
+        {:noreply, {:busy, queue, codel}}
+      _ ->
+        {:noreply, {:busy, queue, codel}}
+    end
+  end
+
+  defp drop_slow(time, timeout, queue) do
+    min_sent = time - timeout
+    match = {{:"$1", :_, :"$2"}}
+    guards = [{:<, :"$1", min_sent}]
+    select_slow = [{match, guards, [{{:"$1", :"$2"}}]}]
+    for {sent, from} <- :ets.select(queue, select_slow) do
+      drop(time - sent, from)
+    end
+    :ets.select_delete(queue, [{match, guards, [true]}])
```

Comment:
I wonder if it would be better to select the full key in the previous pass and then do the deletes by key instead.

Reply:
By better do you mean faster or clearer? I think ETS should be able to optimize this so its faster as its an ordered set so becomes o(n) where n is length to delete, whereas delete each would be o(nlog(m)). However code is likely clearer.